sycl/include/sycl/types.hpp

```diff
@@ -2085,298 +2085,119 @@ __SYCL_RELLOGOP(||)
-  public:                                                                      \
-    using DataType = detail::select_apply_cl_t<                                \
-        bool, __SYCL_GET_CL_TYPE(char, num), __SYCL_GET_CL_TYPE(short, num),   \
-        __SYCL_GET_CL_TYPE(int, num), __SYCL_GET_CL_TYPE(long, num)>;          \
-  };
-
-#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
-#define __SYCL_DECLARE_BYTE_CONVERTER(num)                                     \
-  template <> class BaseCLTypeConverter<std::byte, num> {                      \
-  public:                                                                      \
-    using DataType = __SYCL_GET_CL_TYPE(uchar, num);                           \
-  };
-#endif
-#define __SYCL_DECLARE_HALF_CONVERTER(base, num)                               \
-  template <> class BaseCLTypeConverter<base, num> {                           \
-  public:                                                                      \
-    using DataType = __SYCL_GET_CL_HALF_TYPE(base, num);                       \
-  };
 
-#define __SYCL_DECLARE_SCALAR_SCHAR_CONVERTER                                  \
-  template <> class BaseCLTypeConverter<schar, 1> {                            \
-  public:                                                                      \
-    using DataType = schar;                                                    \
-  };
-
-#define __SYCL_DECLARE_SCALAR_BOOL_CONVERTER                                   \
-  template <> class BaseCLTypeConverter<bool, 1> {                             \
-  public:                                                                      \
-    using DataType = bool;                                                     \
-  };
+// Vectors of size 1 are handled separately and therefore 1 is not included in
+// the check below.
+constexpr bool isValidVectorSize(int N) {
+  return (N >= 2 && N <= 4) || N == 8 || N == 16;
+}
+template <typename T, int N, typename V> struct VecStorage {
+  static_assert(
+      isValidVectorSize(N) || N == 1,
+      "Incorrect number of elements for sycl::vec: only 1, 2, 3, 4, 8 "
+      "or 16 are supported");
+  static_assert(!std::is_same_v<V, void>, "Incorrect data type for sycl::vec");
+};
 
-#if (!defined(_HAS_STD_BYTE) || _HAS_STD_BYTE != 0)
-#define __SYCL_DECLARE_SCALAR_BYTE_CONVERTER                                   \
-  template <> class BaseCLTypeConverter<std::byte, 1> {                        \
-  public:                                                                      \
-    using DataType = __SYCL_GET_SCALAR_CL_TYPE(uchar);                         \
+#ifdef __SYCL_USE_EXT_VECTOR_TYPE__
+template <typename T, int N> struct VecStorageImpl {
+  using DataType = T __attribute__((ext_vector_type(N)));
+};
+#else
+// When ext_vector_type is not available, we rely on cl_* types from CL/OpenCL.h
+// to represent vec storage.
+// FIXME: there is no direct include of CL/OpenCL.h in this file
```

Comment:
> > is it ok to directly include CL/OpenCL.h into this file? Or should we create a proxy for that include to limit number of direct references the header has in our codebase?
> 
> A proxy would be nice to unify the work for removing the headers if we ever need it, but I don't know if it is worth it if we only know of this use-case.

It turned out that we already have such proxy header. Re-used it in `types.hpp`

> > Should we include the file, but import its content into an internal namespace to avoid polluting global namespace?
> 
> That definitely sounds like a good idea.

I suggest we do this in a separate PR to avoid too many changes (and mixing unrelated changes together)